Sources/CollectionView.swift

```diff
@@ -179,43 +185,58 @@ open class CollectionView: UIScrollView {
     }
     let contentOffsetDiff = contentOffset - oldContentOffset
 
-    let newVisibleIndexes = provider.visibleIndexes(activeFrame: activeFrame)
-
-    let newVisibleIdentifiers = Set(newVisibleIndexes.map { index in
-      return newIdentifiersToIndexMap[index]!
-    })
-    let oldVisibleIdentifiers = Set(visibleIndexes.map { index in
-      return identifiersToIndexMap[index]!
-    })
-
-    let deletedVisibleIdentifiers = oldVisibleIdentifiers.subtracting(newVisibleIdentifiers)
-    let insertedVisibleIdentifiers = newVisibleIdentifiers.subtracting(oldVisibleIdentifiers)
-    let existingVisibleIdentifiers = newVisibleIdentifiers.intersection(oldVisibleIdentifiers)
-
-    for identifier in existingVisibleIdentifiers {
-      // move the cell to a different index
-      let oldIndex = identifiersToIndexMap[identifier]!
-      let newIndex = newIdentifiersToIndexMap[identifier]!
-      let cell = visibleCellToIndexMap[oldIndex]!
-
-      newVisibleCellToIndexMap[newIndex] = cell
-      provider.update(view: cell, at: newIndex)
-      cell.currentCollectionPresenter = cell.collectionPresenter ?? provider.presenter(at: newIndex)
-      (cell.currentCollectionPresenter ?? presenter).shift(collectionView: self, delta: contentOffsetDiff, view: cell, at: newIndex, frame: provider.frame(at: newIndex))
-    }
+    let newVisibleIndexesSet = provider.visibleIndexes(activeFrame: activeFrame)
+    let newVisibleIndexes = Array(newVisibleIndexesSet)
 
-    for identifier in deletedVisibleIdentifiers {
-      disappearCell(at: identifiersToIndexMap[identifier]!)
+    let newVisibleIdentifiers = newVisibleIndexes.map { index in
+      return newIdentifiers[index]
+    }
+    let oldVisibleIndexes = Array(visibleIndexes)
+    let oldVisibleIdentifiers = oldVisibleIndexes.map { index in
+      return identifiers[index]
     }
 
-    visibleIndexes = newVisibleIndexes
-    visibleCellToIndexMap = newVisibleCellToIndexMap
-    identifiersToIndexMap = newIdentifiersToIndexMap
+    let diff = oldVisibleIdentifiers.extendedDiff(newVisibleIdentifiers)
+
+    diff.forEach { element in
+      switch element {
+      case .delete(let at):
+        let info = oldVisibleIdentifiers[at].info
+        disappearCell(info: info)
+      case .insert(let at):
+        let index = newVisibleIdentifiers[at].info.1
+        let newCell = appearCell(at: index)
+        newIdentifiers[index].info = (newCell, index)
+      case let .move(from, to):
+        let oldInfo = oldVisibleIdentifiers[from].info
```

Comment:
Basically I wanted to make the provider return an array of visible indexes that indicates the view order. Then in this case, we can move the cell to its correct index inside the superview's subviews array. We can then get rid of the hacky `insert(cell:)` method